migration_toolkit_for_containers/mtc-migrating-vms.adoc

```diff
@@ -0,0 +1,82 @@
+:_mod-docs-content-type: ASSEMBLY
+[id="mtc-migrating-vms"]
+= Migrating virtual machine storage
+include::_attributes/common-attributes.adoc[]
+:context: mtc-migrating-vms
+
+toc::[]
+
+{mtc-full} ({mtc-short}) 1.8.4 supports storage migration for KubeVirt virtual machines (VMs).
+
+The migration can occur offline, when the VM is turned off, or online, when the VM is running.
+
+The main use case for this storage migration for KubeVirt VMs is migration from one storage class to another. For example, if you are rebalancing between different storage providers, or if there is new storage available that is better suited to the workload running inside the VM.
+
+== Supported persistent volume actions
+
+.Supported persistent volume actions
+[width="100%",cols="31%,33%,36%",options="header",]
+|===
+|Action
+|Supported
+|Description
+
+|Copy
+|Yes
+a|
+. Create a new persistent volume (PV) in the same namespace.
+. Copy data from the source PV to the target PV, and modify the VM definition to point to the new PV.
+** If the `liveMigrate` flag is set, the VM migrates live.
+** If the `liveMigrate` flag is not set, the VM shuts down, and the source PV contents are copied to the target PV and the VM is restarted.
+
+|Move
+|No
+|This action is not supported.
+|===
+
+== Preparing for migration
+
+To prepare for migration, deploy mig-controller and Velero, and then create the following `Mig`' resources on the cluster where the migration controller is running to prepare for migration:
+
+//. Deploy link:https://github.com/kubevirt/kubevirt/[KubeVirt]. In order to support storage live migration, you need to deploy KubeVirt version 1.3.0 or later. Earlier versions do not support storage live migration. You also need to configure KubeVirt to enable storage live migration according to the link:https://kubevirt.io/user-guide/storage/volume_migration/[Volume Migration].
+
+. Deploy xref:../virt/install/installing-virt.adoc#installing-virt-operator_installing-virt[{VirtProductName}]. In order to support storage live migration, you need to deploy {VirtProductName} version {VirtVersion}. Earlier versions do not support storage live migration. You also need to configure KubeVirt to enable storage live migration according to the xref:../virt/live_migration/virt-configuring-live-migration.adoc#virt-configuring-live-migration-limits_virt-configuring-live-migration[Configuring live migration].
```

Comment:
```suggestion
. Install xref:../virt/install/installing-virt.adoc#installing-virt-operator_installing-virt[{VirtProductName} Operator]. In order to support storage live migration, you need to deploy {VirtProductName} version {VirtVersion}. Earlier versions do not support storage live migration. You also need to configure KubeVirt to enable storage live migration according to the xref:../virt/live_migration/virt-configuring-live-migration.adoc#virt-configuring-live-migration-limits_virt-configuring-live-migration[Configuring live migration].

+
In {VirtProductName} 4.17.0 not all the required feature gates are enabled. 
As a result, you have to enable the feature gate before being able to use the storage live migration feature.
To work around this, you can use a JSON patch:
+
[source,terminal]
----
$ oc annotate --overwrite -n openshift-cnv hco kubevirt-hyperconverged kubevirt.kubevirt.io/jsonpatch='[ {"op": "add", "path": "/spec/configuration/developerConfiguration/featureGates/-", "value": "VolumesUpdateStrategy"}, {"op": "add", "path": "/spec/configuration/developerConfiguration/featureGates/-", "value": "VolumeMigration"} ]'
----
+
[WARNING] 
====
This should not be done in a production cluster, as it will take the  cluster out of support, as Red Hat does not support clusters with this annotation. A warning will be fired by the creation of this annotation.
====
```